### ISSUE TITLE
Default Severity

### ISSUE THREAD

**Issue by LukeSamkharadze** [eslint/eslint#14679]
**The version of ESLint you are using.**
7.28.0

**The problem you want to solve.**
Now both real code errors and ESLint errors appear at the same level in IDEs with red squiggles.
I can see an option to change all error severity to warn.
It would be nice to have yellow squiggles on ESLint ones (Now we have to manually list every rule and change its severity to `warn`)

**Your take on the correct solution to the problem.**
```
{
  severity: 'warn',
  ...
}
```

**Comment by LukeSamkharadze** [comment]
https://github.com/eslint/eslint/issues/10438

As I can see In the past issue was opened about same topic, but It was marked as a question, so I am opening this one as a feature request

**Comment by nzakas** [comment]
How would you envision this working? Let’s say you set `severity: error` in the config, how do rules know to use that?

**Comment by LukeSamkharadze** [comment]
Without affecting third-party config rules, ESLint itself would overwrite rules severity

A better word would be `maxSeverity`
```
{
  // Default would be 'error'
  maxSeverity: 'warn'
}
```

Or another way is that we could try mapping severities
```
{
  severity: {
    error: 'warn',
    warn: 'off',
    off: 'warn',
  }
}
```

These both ways would affect any config in `extends` array, for example, if you have
```
{
  maxSeverity: 'warn',
  extends: [
    'airbnb',
    'plugin:prettier/recommended'
  ]
}
```

It is debatable if it would also affect our custom rules, but as I think better way would be no
```
{
  maxSeverity: 'warn',
  rules: {
    no-empty: 'error'
  }
}
```

**Comment by nzakas** [comment]
Sorry, I think I was unclear. I’m asking how will users know when a rule is using the default severity? All rules require a severity to be specified, so what will be different when a rule is supposed to use the default severity?

and I don’t think max severity makes sense. Default severity is easier to understand.

**Comment by LukeSamkharadze** [comment]
Ah, We could have
```
rules: {
  no-empty: 'default',
  no-var: 4
}
```

-----

But the thing is, for example in [**Airbnb-config**](https://www.npmjs.com/package/eslint-config-airbnb) or any other config file you use, they have to change every rule to **default** or they may not at all. In this case, `default` won't work. We could have mapping severities or `severity: 'warn'` that will change all rules severity, but not the ones you specified in `rules` object.

The main problem I wanted to solve is to control third-party configuration that your config extends with only 1 field. Even if they have `error` severity instead of `default`

Imagine you are extending these rules
```
extends: [
  'airbnb-typescript/base',        
  'eslint:recommended',        
  'plugin:@typescript-eslint/eslint-recommended',        
  'plugin:@typescript-eslint/recommended',        
  'plugin:@typescript-eslint/recommended-requiring-type-checking',        
  'plugin:@angular-eslint/recommended',        
  'plugin:@angular-eslint/recommended--extra',
  'plugin:prettier/recommended'
]
```

While coding and seeing errors in every line you type, is not very development-friendly...

With the option to control the `severity` of all config files, you can turn ESLint **dictatorship** into some kind of helper tool that reminds you to use a better way of fixing problems

**Stylelint** has a similar feature that makes working so much comfortable -> [defaultSeverity](https://stylelint.io/user-guide/configure/#defaultseverity)

**Comment by nzakas** [comment]
Thanks, I think I get it now. This seems interesting enough to explore in some further detail for implementing in the new config system (we aren’t making any further changes to the eslintrc format).

The TSLint model of having a defaultSeverity that is used whenever the severity is set as “default” for a rule makes sense to me. That won’t let you override other statically-set rule severities but it does give you more options.

Stylelint is a bit different because you don’t have to set a severity on rules, which is basically the equivalent of setting “default”.

However, neither of these addresses your use case:

> The main problem I wanted to solve is to control third-party configuration that your config extends with only 1 field. Even if they have error severity instead of default

In this case what you are really asking for is a way to *override* all rule severities at once even if they are explicitly set. This might be a bit too complicated, though, because we don’t really distinguish between configs when merging. I’m not sure there’s a way to do that without making things overly complex.

**Comment by LukeSamkharadze** [comment]
Thanks for following up and amazing info 

Let's discuss previous options and few new ones and try to come up with the best one

### Severity ⛔
I don't think this is a good option as it can also override rules with `off`. Some plugins override previous ones and turn some rules off to avoid conflicts. (Solution might be to ignore rules with `off`)
```JS
{
  severity: 'warn', // Will override all rule's severity while merging extends array
}
```

### MaxSeverity ⛔
It will be quite good for strict and visual config files
```JS
{
  maxSeverity: 'warn', // Will downgrade all rule's severity while merging extends array
}
```

### Mapping Severity ❔
Can easily replicate **Severity**'s and **MaxSeverity's** behaviour with additional capabilities
```JS
{
  // Will override all rule's severity while merging extends array according to this map
  severity: {
    error: 'warn',
    warn: 'off',
    off: 'warn',
  }
}
```

New Ideas:

### MaxSeverity next to extending config ❔
Simple one to implement. 
Use cases: Tone down strict configs
```JS
{
  extends: [
    [ 'airbnb-typescript/base', 'warn' ], // This option will act as MaxSeverity, recommended for strict or visual configs
    [ 'eslint:recommended', 'error'], // This will be default
    'plugin:@typescript-eslint/eslint-recommended',      
    'plugin:@typescript-eslint/recommended',        
    [ 'plugin:@typescript-eslint/recommended-requiring-type-checking', 'warn' ],        
    'plugin:@angular-eslint/recommended',        
    [ 'plugin:@angular-eslint/recommended--extra', 'warn' ],
    [ 'plugin:prettier/recommended', 'warn' ]
  ]
}
```

### Mapping & MaxSeverity next to extending config ❔
Best one
```JS
{
  // Global severity that will act if we don't use array for config
  severity: {
    warn: 'off'
  },
  extends: [
    'plugin:@typescript-eslint/recommended-requiring-type-checking', // <-- Global Severity will act
    [ 'airbnb-typescript/base', 'warn' ], // This 'warn' is a MaxSeverity and will ignore global Severity
    [ 
      'eslint:recommended',
      { warn: 'error' } // If its object, treat as Mapping Severity and ignore global Severity
    ]
  ]
}
```
```JS
// If you want to have MaxSeverity at the global level
{
  severity: {
    error: 'warn'
  },
  extends: [
    ...
  ]
}
```

These ones will give us the ability to flexibly override third-party config files.

If you think this last way introduces unnecessary complexity, we can think about the **Simple one**, **Only Mapping** at global level or adding `default` and `defaultSeverity` options.

<!---
-----
### For the Default

This is quite a separate topic, we can also add new severity called `default` and work with `defaultSeverity`.

In this case, next to extending config we would have defaultSeverity

```JS
{
  // We could also have default mapping and ability specify for each third party config
   severity: {
    default: 'warn',
    error: 'warn',
    warn: 'off',
    off: 'warn',
  },
  extends: [
    [ 'airbnb-typescript/base', 'warn' ], // This 'warn' will be DefaultSeverity instead of MaxSeverity
    [ 
      'eslint:recommended',
      { warn: 'error' } // If its object treat as Mapping Severity
    ],
    [ 'plugin:prettier/recommended', 'warn' ]
  ]
}
```
---!>

**Comment by nzakas** [comment]
Thanks for mapping those out...I'm not sure any of those is simpler than the others implementation-wise. The TSLint approach of `defaultSeverity` and using `default` is actually the easiest to deal with.

Also, keep in mind that we are no longer making changes to the current config system, so any changes need to take into account how the [new config system](https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification) will work, which currently doesn't have the notion of "extends".

At this point, we need to wait for some more input from the team to decide how to proceed here.

**Comment by mdjermanovic** [comment]
> Now both real code errors and ESLint errors appear at the same level in IDEs with red squiggles.
> I can see an option to change all error severity to warn.
> It would be nice to have yellow squiggles on ESLint ones (Now we have to manually list every rule and change its severity to `warn`)

If the goal is to differentiate errors coming from ESLint and those from other sources in IDEs, that sounds like something IDE-ESLint integrations should do.

[vscode-eslint](https://github.com/microsoft/vscode-eslint) provides this feature with `eslint.rules.customizations` option:

> - `eslint.rules.customizations` (since 2.1.20): force rules to report a different severity within VS Code compared to the project's true ESLint configuration. Contains two properties:
>   - `"rule`": Select on rules with names that match, factoring in asterisks as wildcards: `{ "rule": "no-*", "severity": "warn" }`
>     - Prefix the name with a `"!"` to target all rules that _don't_ match the name: `{ "rule": "!no-*", "severity": "info" }`
>   - `"severity"`: Sets a new severity for matched rule(s), `"downgrade"`s them to a lower severity, `"upgrade"`s them to a higher severity, or `"default"`s them to their original severity
> 
>   In this example, all rules are overridden to warnings:
> 
>   ```json
>   "eslint.rules.customizations": [
>     { "rule": "*", "severity": "warn" }
>   ]
>   ```
> 
>   In this example, `no-` rules are informative, other rules are downgraded, and `"radix"` is reset to default:
> 
>   ```json
>   "eslint.rules.customizations": [
>     { "rule": "no-*", "severity": "info" },
>     { "rule": "!no-*", "severity": "downgrade" },
>     { "rule": "radix", "severity": "default" }
>   ]
>   ```

**Comment by LukeSamkharadze** [comment]
@mdjermanovic, 

Thanks for the reply, that would greatly improve the situation but will have no luck with pre-commit hooks that check code for ESLint errors

**Comment by mdjermanovic** [comment]
> but will have no luck with pre-commit hooks that check code for ESLint errors

Can you elaborate more on this use case, what is the desired behavior?

**Comment by LukeSamkharadze** [comment]
Fully change linting severity to **warning**, So if I type `npx eslint .` it would just show warnings instead of errors.

This will solve errors on pre-commit hooks & annoying red squiggles at the same time.

**Comment by nzakas** [comment]
I still kind of like the defaultSeverity approach of TSLint. What do others think?

**Comment by nzakas** [comment]
Oops, forgot to assign to myself.

**Comment by nzakas** [comment]
@eslint/eslint-tsc thoughts on adding a defaultSeverity?

**Comment by btmills** [comment]
Is there still a motivating use case for a default severity? `eslint.rules.customizations` takes care of the editor integration case, and scripting can be solved with `|| exit 0`.

**Comment by LukeSamkharadze** [comment]
@btmills could you provide more info on `eslint.rules.customizations`?

Scripting should be fixale per rule `|| exit 0` will skip eslint `errors` too

**Comment by nzakas** [comment]
> Is there still a motivating use case for a default severity? 

In my mind, this would allow shareable configs to set their rule severities to “default”, which would make it easier for consumers to integrate into their own configs where, for example, all rules might be set to warn.

**Comment by btmills** [comment]
The shareable config use case makes sense :+1:

**Comment by bmish** [comment]
Just trying to make sure I understand this proposal. I'm assuming the current proposal is:

* Shareable configs can set `'some-cool-rule': 'default'`
* End-users can set `defaultSeverity: 'error'` to choose the severity of any rule that specifies `default`

So far, I have the following concerns:

1. Wouldn't it create a huge amount of work for shareable configs to have to update all of their rules to use `default` instead of `error`/`warn`? This proposal seems like it's solving the problem by pushing work onto all of the thousands of shareable ESLint configs, instead of the alternative solution of just letting the end-user apply severity overrides when needed.
2. How would shareable configs decide when `default` is appropriate to use vs. more specific settings like `error`/`warn`? Seems like there would inevitably be disputes here, as some config owners would still want to choose specific severities, but some end-users might want them to specify `default`.
3. What about shareable configs that never get around to adopting `default`? Seems like we would still need to have the severity override solution given that we'd never get to full adoption of `default` anyway.
4. If we end up in a situation where configs just use `default` for everything, than that potentially conveys less information than today where config owners can use `error` / `warn` to distinguish stronger vs. weaker rule suggestions. Not the end of the world, but it's another downside to the severity override solution where end-users could override `error` / `warn` rules differently.

**Comment by nzakas** [comment]
> Wouldn't it create a huge amount of work for shareable configs to have to update all of their rules to use default instead of error/warn?

It’s an option shareable configs can choose to use or not. It’s not required.

> How would shareable configs decide when default is appropriate to use vs. more specific settings like error/warn?

Again, that’s up to the author. The shareable config use case is just the most obvious to me but I doubt it’s the only one. I could also imagine people using this in their own configs to easily switch severity levels depending on the environment.

> What about shareable configs that never get around to adopting default? 

See previous answers. 

> If we end up in a situation where configs just use default for everything, then that potentially conveys less information than today where config owners can use error / warn to distinguish stronger vs. weaker rule suggestions.

Again, config owners can decide to do whatever they want. Undoubtedly, configs that are used primarily within a company will likely not use “default” whereas configs intended for public consumption may prefer “default”.

> So a severity override solution that end-users could configure as needed makes much more sense to me at this point than an ecosystem-wide default severity migration.

First, this isn’t an either-or situation. I think you misunderstood my example use case as being the definitive one and perhaps read in a lot more intent than I have.

What do you mean by a severity override? How do you envision that working?

**Comment by bmish** [comment]
The severity override alternative I was describing would look like this (@LukeSamkharadze originally suggested it in a comment above):

Turn warnings off:

```
{
  severity: {
    warn: 'off'
  }
}
```

Upgrade warnings to errors:

```
{
  severity: {
    warn: 'error'
  }
}
```

Downgrade errors to warnings and turn off errors:

```
{
  severity: {
    error: 'warn',
    warn: 'off'
  }
}
```

This seems significantly better to me compared to introducing `default` severity, as it puts the end-user in control of severities instead of config owners. Does the `default` severity solution have any advantage over this? From what I understand, it has a lot of downsides.

**Comment by ljharb** [comment]
Config owners need that too though. Generally speaking anything that can go in a top level config is something a shared config would want to abstract.

**Comment by nzakas** [comment]
@bmish I just wanted to double check. I’m just not a fan of that type of mapping both because it’s overloading the meaning of words (“warn” means “off” now?) and because it’s not clear what happens if I have an override for “warn” but then explicitly set a rule to “warn” (did you intend to use the override or not?). 

> Does the default severity solution have any advantage over this? From what I understand, it has a lot of downsides.

Sorry, what downsides are you referring to here? As already mentioned, this is strictly opt-in for shared config authors.

To me, “default” clearly indicates that the end user is making the severity decision. (Keep in mind, defaultSeverity was [implemented in StyleLint](https://stylelint.io/user-guide/configure/) so there is prior art here.)

@ljharb can you clarify what you mean? Are you saying you’d like both a default severity and an override?

**Comment by ljharb** [comment]
@nzakas im saying that regardless of what the default is, i want to be able to override it inside my shared config.

Separately, and secondarily, i prefer stricter defaults :-)

**Comment by LukeSamkharadze** [comment]
My only problem with extending large configs was that I wanted all of the `eslint` problems to be marked as warnings and then manually mark rules that I personally thought them as being errors.
Having top level overriding severities would solve the problem, but I think it would be better to think of more robust solution.

Here is why mapping severities per extended config could be better. Take 4 configs as an example:

1. Config `A`  has logically assigned severity to every issue
3. Config `B` thinks that all of his problems should be marked as an errors
5. Config `C` uses warn for majority of the rules but in our opinion they should be errors
7. Config `D` has nonsenses as a warn

Global mapping severities won't help in this case, since we still have to manually override majority of the rules ourselves, so what shoud we do?

```JS
{
  extends: [
    'A',
    [ 'B', { error: 'warn' } ], // Will map B's errors to warns 
    [ 'C', { warn: 'error' } ], // Will map C's warns to errors
    [ 'D', { warn: 'off' } ] // Will map D's warns to off
  ],
  rules: {
    "B/something-not-important": "warn", // Not so important warn should still be warn
    "C/something-important": "error" // Important rule in C that should be error
  }
}
```

I might be **nitpicking** with this example, but at least top level overriding should still be the choice for end-users instead of relying on their config **provider** to use default logically or to us it at all

**Comment by nzakas** [comment]
> but at least top level overriding still should be the choice for the end-users instead of them relying on their config provider to use default seveirty logically or to us it at all

This covers just one side of the equation: when a shared config provider wants to allow consumers to override severities. The other side also exists: where shared config providers explicitly do not want to allow severities to be changed. We can think of this as the open source model vs the internal model, where someone with a shared config inside of a company wants to enforce their severities across the board.

The defaultSeverity approach enables not use cases while the severity mapping approach only enables one.

**Comment by ljharb** [comment]
How is that ever preventable? I can always use a JS config and transform/override the shared config any way i want. As a consumer, i can even use npm overrides to replace it with my own package.

I don’t think consumers can ever truly be restricted by shared configs, nor do i think that’s a particularly valid use case - inside a company, you’d do that with code-owned CI checks.

**Comment by nzakas** [comment]
Not everyone wants to go through the trouble to create an intermediate config file that adjusts shared config settings, especially if they need to do that for multiple shared configs, or shared configs that extend other shared configs (which leaves you unable to make such wholesale changes). In a company setting, that’s even less likely to happen. I’m not too concerned about what people do on their personal projects.

But I’d like to pop the stack here. I’m fine with closing this issue and not making any changes if there’s no agreement on how to move forward. I think defaultSeverity is a good idea with utility and would be fairly easy to implement, and I can’t say the same about the severity mapping proposal, which just seems way too confusing to me and would be difficult to implement. 

Note that in the new config system, shareable config authors could also export a function where they allow end users to pass in parameters and produce a config based on that (because you’ll be able to manually import and use any package you want in any way you want), so perhaps we can just wait and see if that also solves the cases mentioned.

**Comment by ljharb** [comment]
That wouldn’t be consumable by a non-js config tho :-/

**Comment by LukeSamkharadze** [comment]
It's not that I'm against  `defaultSeverity`. I was just trying to find solution that would give end-users much flexibility.

Introducing `defaultSeverity` rather than closing the issue would be better.

**Comment by bmish** [comment]
I guess I don't feel like my concerns about the downsides of the default severity solution have been adequately addressed: https://github.com/eslint/eslint/issues/14679#issuecomment-1003578249.

Separately, I would ask, has this been issue been a long-standing pain-point / demand that justifies the added complexity of any of these solutions?

**Comment by nzakas** [comment]
@bmish I addressed each of your concerns in [this comment](https://github.com/eslint/eslint/issues/14679#issuecomment-1005265857). As I said, it seems like your primary concerns are all around the assumption that shareable configs would be required to use `default`, but since they wouldn't be, it seems like most of those concerns go away. Am I incorrect about that?

And yes, there has been a longstanding complaint that there isn't an easy way to change the severity of a bunch of rules at once. I've actually looked at implementing `defaultSeverity` and it's not complex at all (in the new config system). There's a finalization step when configs are calculated that would allow us to very easily swap in a different severity for rules with `default`.

But as I said, I think at this point it's probably most prudent to wait until the new config system can be played with because we do have other options for how to achieve the same thing.

**Comment by btmills** [comment]
As I was catching up with this thread since my last comment, I think I’ve come to a similar conclusion that we should hold off making any change now. In the current config system, coming up with the right API for this is difficult. Meanwhile, in the upcoming confit format, all variants of this use case are relatively straightforward and can be done in a few lines of code.

**Comment by melink14** [comment]
If there's a new config format coming then waiting makes sense but I'm definitely aligned with bmish on the idea that a default that shareable configs have to use is only half the solution people have been asking for over the years.

I don't think `defaultSeverity` solves the problem of "easy way to change the severity of a bunch of rules at once" as long as shareable configs have to do anything to make it happen. So any solution which ends up being "up to the shareable config authors" I think would be a half solution only. (though still better than no solution!)

Thanks for pushing this topic on this thread, it was heartening to read recent constructive discussion on an issue I was researching!

**Comment by Jackman3005** [comment]
For those of us working in WebStorm and looking to remap warn and error from ESLint to new severity levels in WebStorm, I created a request [here](https://youtrack.jetbrains.com/issue/WEB-66035/Allow-overriding-individual-ESLint-severity-levels-to-desired-inspection-severity). Note that WebStorm does already support remapping both ESLint error and warn to a single new severity level in WebStorm, which may already be what you want. However, I would like to be able to map warn and error separately, which the request asks for. Would be great to get a thumbs up on that request if you would like it as well!

**Comment by marcospgp** [comment]
arriving from google and seeing this issue as open, is there a way to override all rules being set to error by default when enabling recommended rulesets?
